lib/media/streaming_engine.js

```diff
@@ -1116,6 +1116,7 @@ shaka.media.StreamingEngine = class {
       const ref = mediaState.segmentIterator.next().value;
       if (ref == null) {
         shaka.log.warning(logPrefix, 'cannot find segment', 'endTime:', time);
+        mediaState.segmentIterator = null;
```

Comment:
If we don't clear the mediaSegment.segmentIterator downloading of segments for this stream will stop because mediaState.segmentIterator.current() is null, causing a period of buffering until we fall out of the segment availability window.

Reply:
Does segment iterator not recover after new segments are added?

It looks like `currentPosition_` in `SegmentIterator` would advance to an invalid position after the end of the segments are reached, but that position would no longer be invalid after new segments are added.

Because `MetaSegmentIndex` is built to concatenate several independent indexes, it's possible that the position-based iterator would have produced these bad results prior to your fixes in the generation & eviction logic in lib/dash/.  But if the underlying indexes are correctly maintained, the iterator shouldn't need to be set to null here.

Can you check to see if this change in StreamingEngine is still necessary with the dash fixes?

Reply:
I found this issue after making the dash_parser and segment_template changes. The SegmentIterator does not recover.

Tracing the code.. 

When segmentIndex.getIteratorForTime(time) is called with a time that is later than the last reference in the segment index, SegmentIterator.currentPosition_ is set to -1

When SegmentIterator.next() is called, currentPosition_ advances to 0 and will return null if the position 0 segmentReference has been evicted, which will usually be the case for a live streams. It seems like there would be a real mess if the position 0 reference had not been evicted.

Subsequent  calls to SegmentInterator.current() will return null, and it is stuck.

Reply:
Oh, that's awful.  Do you see a way to fix the iterator instead of resetting it?  If not, can you file a new issue about this and link to it from a code comment at this location so that someone else can follow up later?

Reply:
I think the safest thing to do is to make getIteratorForTime(time) return null if it cannot locate a reference for the requested time, and add checks for that condition in the streaming engine code that calls it.

Reply:
@joeyparrish  can you opine on my suggestion to make getIteratorForTime(time) return null when it cannot locate a reference for the requested time. This will require addition of null checks in the 4 places this function is called.

Reply:
I think there are two important patterns to consider:

1. StreamingEngine hits the end of the iterator, and then more segments are added
2. StreamingEngine hits the end of the iterator, and switches streams/indexes, and **then** more segments are added

In both cases, current() and next() should allow StreamingEngine to continue at the correct time.

I believe changing getIteratorForTime() would only address the second case, because it is only called when StreamingEngine switches streams.  I think we can fix the iterator itself to remember its place, and I would prefer that solution if it can be solved fairly easily.

The good news is that I think I can solve this part.  It's relatively easy to write unit tests for these calling patterns and set expectations on SegmentIterator without involving the complexity of StreamingEngine.  I will try to write tests for these two patterns and then work on fixing the position-tracking in SegmentIterator.  If I'm successful, I propose you revert this part of your change and we just merge your fixes for DashParser in this PR.

What do you think?

Reply:
I got stuck on fixing SegmentIterator because I felt that somehow, it was not doing its job.  Now that I've spent some time trying to fix it, I think I was wrong, and you were right.

The first case I described already works.  The second case I described can be made to work, but it's not perfect.

(As an aside, neither of the cases I described is actually tested, so I will at least add test cases for these.)

For the second case, I can trivially set the position to `length - 1` instead of `-1`, which means iteration will pick up with the first new segment added.

However, if `getIteratorForTime()` is called with a time in the future, that time is not necessarily the very next segment.  For example, if I have times 0 - 100 in the index, and someone calls `getIteratorForTime(800)`, it would not be appropriate for `current()` to suddenly return a segment at time 101 after I add 101 - 200 to the index.

So I think it does make sense for StreamingEngine to ask for an iterator for its desired time again until that time exists in the index.  I'm sorry that it took me so long to catch up to what you have known all along.  Thank you for being so patient with me!

Reply:
I know I wrote a ton here, so to be completely clear, I think you are right that getIteratorForTime() should return null in cases where that time is not in the index.  Thanks!

Reply:
Here is the new test case I added for SegmentIterator: https://github.com/google/shaka-player/commit/53173c5a235ba74b27d03b7003dbc94312045531